solver/src/paraswap_solver.rs

```diff
@@ -1 +1,383 @@
+use std::sync::Arc;
+
+use anyhow::{anyhow, ensure, Result};
+use contracts::{GPv2Settlement, ERC20};
+use ethcontract::{Bytes, H160, U256};
+use maplit::hashmap;
+use shared::{conversions::U256Ext, token_info::TokenInfoFetching};
+
+use crate::{
+    encoding::EncodedInteraction,
+    interactions::Erc20ApproveInteraction,
+    liquidity::LimitOrder,
+    paraswap_solver::api::{PriceQuery, Side, TransactionBuilderQuery},
+    settlement::{Interaction, Settlement},
+    single_order_solver::SingleOrderSolving,
+};
+
+use self::api::{DefaultParaswapApi, ParaswapApi, PriceResponse, TransactionBuilderResponse};
+
 mod api;
+
+const REFERRER: &str = "GPv2";
+
+/// A GPv2 solver that matches GP orders to direct ParaSwap swaps.
+struct ParaswapSolver<F> {
+    settlement_contract: GPv2Settlement,
+    token_info: Arc<dyn TokenInfoFetching>,
+    allowance_fetcher: F,
+    client: Box<dyn ParaswapApi + Send + Sync>,
+}
+
+impl ParaswapSolver<H160> {
+    #![allow(dead_code)]
+    pub fn new(
+        settlement_contract: GPv2Settlement,
+        token_info: Arc<dyn TokenInfoFetching>,
+    ) -> Self {
+        let allowance_fetcher = settlement_contract.address();
```

Comment:
neat! This could maybe also be added to the 1Inch solver eventually :smile:.

One idea (but feel free to ignore, I think it looks good like this as well) would be to instead:
```
trait AllowanceFetching: Send + Sync {
    async fn existing_allowance(&self, token: &H160, spender: &H160) -> result<U256>;
}
impl AllowanceFetching for GPv2Settlement { ... }
```

This way, you wouldn't need a `web3` instance at all to mock `AllowanceFetching` (as there would be no need to construct a `ERC20` instance to pass to the mock fetching) and make it easier to set mock parameter assertions as `H160: Eq`.

Reply:
Agree that this interface would be more appropriate, however we still need the `ERC20` instance to create the `Erc20ApproveInteraction` approve interaction (so we would still need a web3 instance). Unless we refactor `Erc20ApproveInteraction` to take the ERC20 address and encode the calldata with a dummy web3 (which I believe should work).

Reply:
Makes sense, Lets leave it as is then.